optimizer/plan/alternatives.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"fmt"
+)
+
+// Alternatives returns multiple alternative plans that
+// can be picked base on its cost.
+func Alternatives(p Plan) []Plan {
+	var plans []Plan
+	switch x := p.(type) {
+	case nil:
+	case *TableScan:
+		plans = tableScanAlternatives(x)
+	case WithSrcPlan:
+		plans = planWithSrcAlternatives(x)
+	default:
+		panic(fmt.Sprintf("unknown plan %T", p))
```

Comment:
I think use panic is better when it is definitely a program bug than runtime error.